### DISCUSSION TITLE
Outsource Position Controller - Timing Problems

### DISCUSSION THREAD

**Discussion by clubbymaty** [bitcraze/discussions#301] [Q&A] [answered]
Hi there,

for the context: I outsourced the Position Controller of the CF to a Python script on a PC and using the Logging feature to get sensor data and the Commander to send the respective/ calculated Attitude Setpoints with the CF Radio back to the CF. The Synccrazyflie method, which is used in all examples and so is the way I am going here so far, lies in a Control-Loop and gets executed in every iteration of the loop - which reads out the sensors (Simplelog Example) and writes the Setpoints (Commander) which where calculated in the previous iteration.

Using the Synccrazyflie, as in the Examples, I am experiencing a really high delay,  so the execution of Synccrazyflie is very time consuming. So my question at this point is if this delay/ execution time (600ms) is a regular delay based on the communication or is that so high because my PC is not fast enough. **Is there, in terms of timing, a better way to do this?**

I also tried to put the Controller-Loop inside the Synccrazyflie Method but because there only can be 16 Iterations its not useful for my case - the timing of course is a lot better.

best regards

**Comment by krichardsson** [comment, marked as answer]
My understanding is that you are doing something like this:

``` python
while(True):
  # Do stuff
  with SyncCrazyflie(URI) as scf:
    # Get log values using scf
 # Do more stuff
```

That is, you are using `with SyncCrazyflie(URI) as scf:` inside an outer loop?

`with SyncCrazyflie(URI)` is doing a full connect, downloading TOCs and other stuff which takes a long time.
The typical usage is to connect once and then do what you want to after that, so if you are using the `SyncCrazyflie` class your working loop should go inside the SyncCrazyflie scope.
 
The `SyncCrazyflie` was designed to make it easy to write small synchronous scripts, even though the actual communication with the Crazyflie is asynchronous. If you need a more asynchronous behavior in you code you can check out some other example that does not use the `SyncCrazyflie` class, like https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/logging/basiclog.py

You can also mix the two paradigms, my guess is that this is what you should do in this case.

``` python
with SyncCrazyflie(URI) as scf:
  cf = scf.cf
  # Set up logs asynchronously, as is basiclog.py 
  while(True):
    # Do stuff based on the log data you got in your log callback
```

**Comment by clubbymaty** [comment]
Worked with that. I just put the `Logconfig` into the `with SyncCrazyflie` scope and done the `while loop `afterwards where I called the `async log function`.

biggest problem was that the values didn't change over time, been all the same (just the first which was measured) - had to add logconf.delete() at the end of the async log function:

```
def simple_log_async(scf, logconf):
    cf = scf.cf
    cf.log.add_config(logconf)
    logconf.data_received_cb.add_callback(log_stab_callback)
    logconf.start()
    time.sleep(0.04)
    logconf.stop()
    logconf.delete()
```

Thank you!

**Reply by ataffanel** [reply]
The way to use the async log API is to setup a logblock, launch it, and your callback function will be called at regular interval (depending of the logconf period setting). You should not start and stop a logconf so quick, start it at connection and stop it when new values are not needed anymore.

A good example of what the structure of such program can look like is in the autonomousSequence example: the function to setup the logging is [called before](https://github.com/bitcraze/crazyflie-lib-python/blob/8d82a769952d1b49de0d6779a2b59ba94036151a/examples/autonomy/autonomousSequence.py#L150) the function that runs the controlling ```for``` loop, and this function is [setting up and starting a log block](https://github.com/bitcraze/crazyflie-lib-python/blob/8d82a769952d1b49de0d6779a2b59ba94036151a/examples/autonomy/autonomousSequence.py#L116-L124). There is no stopping and deleting the block in this example since it will all be cleanned up when disconnecting.

**Reply by clubbymaty** [reply]
ok, looks like that now and runs also as before, but is there a possibility to reduce the time.sleep? if I go under the 4ms the callback doesn't work anymore.

```
def log_stab_callback(timestamp, data, logconf):
    log_stab_callback.x = data["kalman.stateX"]
    log_stab_callback.y = data["kalman.stateY"]
    log_stab_callback.z = data["kalman.stateZ"]
    log_stab_callback.yaw = data["stateEstimate.yaw"]
    log_stab_callback.vx = data["stateEstimateZ.vx"]
    log_stab_callback.vy = data["stateEstimateZ.vy"]
    log_stab_callback.vz = data["stateEstimateZ.vz"]
    log_stab_callback.yawrate = data["stateEstimateZ.rateYaw"]
```

```
def start_log(scf):
    lg_stab = LogConfig(name='Stabilizer', period_in_ms=10)
    lg_stab.add_variable('kalman.stateX', 'float')
    lg_stab.add_variable('kalman.stateY', 'float')
    lg_stab.add_variable('kalman.stateZ', 'float')
    lg_stab.add_variable('stateEstimate.yaw', 'float')
    lg_stab.add_variable('stateEstimateZ.vx', 'int16_t')
    lg_stab.add_variable('stateEstimateZ.vy', 'int16_t')
    lg_stab.add_variable('stateEstimateZ.vz', 'int16_t')
    lg_stab.add_variable('stateEstimateZ.rateYaw', 'int16_t')
    scf.cf.log.add_config(lg_stab)
    lg_stab.data_received_cb.add_callback(log_stab_callback)
    lg_stab.start()
    time.sleep(0.04)
```

just took the setup of the logconf out from the `with SyncCrazyflie` scope (was not in the control loop) and put it into the function to call the log/ get the values from the callback. with `log_stab_callback.x` to use the value in the controller afterwards.
